specification/containerservices/resource-manager/Microsoft.ContainerService/preview/2018-09-30-preview/openShiftManagedClusters.json

```diff
@@ -0,0 +1,650 @@
+    },
+    "OpenShiftManagedClusterAuthProfile": {
+      "properties": {
+        "identityProviders": {
+          "type": "array",
+          "items" : {
+            "$ref": "#/definitions/OpenShiftManagedClusterIdentityProviders"
+          },
+          "description": "Type of authentication profile to use."
+        }
+      },
+      "description": "OpenShiftManagedClusterAuthProfile defines all possible authentication profiles for the OpenShift cluster."
+    },
+    "OpenShiftManagedClusterProperties": {
+      "properties": {
+        "provisioningState": {
+          "readOnly": true,
+          "type": "string",
+          "description": "The current deployment or provisioning state, which only appears in the response."
+        },
+        "openShiftVersion": {
+          "type": "string",
+          "description": "Version of OpenShift specified when creating the cluster."
+        },
+        "publicHostname": {
+          "type": "string",
+          "description": "Optional user-specified FQDN for OpenShift API server."
+        },
+        "fqdn": {
+          "type": "string",
+          "description": "User-specified FQDN for OpenShift API server loadbalancer internal hostname."
+        },
+        "routerProfiles": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/OpenShiftRouterProfile"
+          },
+          "description": "Configuration for OpenShift router(s)."
+        },
+        "masterPoolProfile": {
+          "$ref": "#/definitions/OpenShiftManagedClusterMasterPoolProfile",
+          "description": "Configuration for OpenShift master VMs."
+        },
+        "agentPoolProfiles": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/OpenShiftManagedClusterAgentPoolProfile"
+          },
+          "description": "Configuration of OpenShift cluster VMs."
+        },
+        "authProfile": {
+          "$ref": "#/definitions/OpenShiftManagedClusterAuthProfile",
+          "description": "Configures OpenShift authentication."
+        }
+      },
+      "required": [
+        "openShiftVersion"
+      ],
+      "description": "Properties of the openshift managed cluster."
+    },
+    "OpenShiftManagedCluster": {
+      "allOf": [
+        {
+          "$ref": "#/definitions/Resource"
+        },
+        {
+          "properties": {
+            "plan" : {
+              "description": "Define the resource plan as required by ARM for billing purposes",
+              "$ref": "#/definitions/PurchasePlan"
+            },
+            "properties": {
+              "description": "Properties of a openshift managed cluster.",
+              "$ref": "#/definitions/OpenShiftManagedClusterProperties",
+              "x-ms-client-flatten": true
+            }
+          }
+        }
+      ],
+      "description": "OpenShift Managed cluster."
+    },
+    "OpenShiftManagedClusterServiceAADIdentityProvider": {
```

Comment:
This is used in `OpenShiftManagedClusterIdentityProviders` in the provider field as `object` we are planning to support more auth in the future

Reply:
But this generated model will not be used in response deserialization?

Reply:
It will be used by the sdk to create a authentication model? Makes sense ?

Reply:
Hi @julienstroheker ,
Yes and no - technically this will work, however the standard approach here is to use polymorphic models. I see for example, that `OpenShiftManagedClusterServiceAADIdentityProvider` has a "kind" property. We generally consider this to be a discriminator (as I suspect this should be a constant value?). If all your intended "IdentityProvider" models will likely have a "kind" property - then I would recommend that you restructure this to use an "OpenShiftManagedClusterServiceIdentityProvider" base class.

Here is an example:
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/stable/2017-12-01/postgresql.json#L1355

You can see the base class "ServerPropertiesForCreate" defines the descriminator "createMode" (in your case this would be "kind") which is then set for each of the children using the "x-ms-discriminator-value" extension. In this example the base class has additional common properties - but that's not required depending on your case.

Here is an additional example:
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/graphrbac/data-plane/stable/1.6/graphrbac.json#L2327

In this case, like yours, the base class only has one available child class at present - which leaves them the option to add more later.

Reply:
Agree with you @annatisch  here, we are working with RedHat on this service and our API have to respect their interface so we are limited for now. This architecture applies for the preview and may change for GA.

You can find the RedHat's interface here : https://github.com/openshift/openshift-azure/blob/master/pkg/api/2018-09-30-preview/api/types.go#L170

Reply:
This should works indeed ! Let me run some tests with it ! Thanks @annatisch

Reply:
@annatisch I just pushed the modification, can check and tell me if it is ok?

Reply:
Thanks @julienstroheker - not quite - You can remove the x-ms-enum extension and it also needs to be marked as required.
Take a look at the docs here:
https://github.com/Azure/autorest/blob/bc0dcde0a05606a1b102c0e08c1798fa4da4dcc8/docs/extensions/readme.md#single-value-enum-as-a-constant

Reply:
What about now ? :)